dataservice/api/biospecimen/manager.py

```diff
@@ -0,0 +1,314 @@
+    """
+    # Extract the parameters that uniquely identify a sample
+    params = _get_sample_identifier(biospecimen)
+    # Add remaining sample attributes
+    params.update(
+        {
+            "participant_id": biospecimen.participant_id,
+            "external_id": biospecimen.external_sample_id,
+            "volume_ul": biospecimen.volume_ul,
+        }
+    )
+    # Set visibility params based on Biospecimen which represents both
+    # sample and containers
+    params.update(
+        _get_visibility_params(biospecimen)
+    )
+    try:
+        sample = SampleSchema(strict=True).load(
+            params, instance=current_sample, partial=True
+        ).data
+    except ValidationError as e:
+        abort(400, 'could not update sample: {}'.format(e.messages))
+
+    return sample
+
+
+def _create_container(biospecimen, sample):
+    """
+    Create Container using specific Biospecimen attributes.
+    Link Container to its associated biospecimen and sample
+    Validate Container
+    """
+    # Extract the parameters that uniquely identify a sample
+    params = _get_container_identifier(biospecimen)
+    # Add remaining sample attributes
+    params.update(
+        {
+            "biospecimen_id": biospecimen.kf_id,
+            "sample_id": sample.kf_id,
+            "volume_ul": biospecimen.volume_ul,
+            "external_id": biospecimen.external_aliquot_id,
+        }
+    )
+    # Set visibility params based on Biospecimen which represents both
+    # sample and containers
+    params.update(
+        _get_visibility_params(biospecimen)
+    )
+    try:
+        container = ContainerSchema(strict=True).load(params).data
+    except ValidationError as e:
+        abort(400, 'could not create container: {}'.format(e.messages))
+
+    return container
+
+
+def _update_container(current_container, biospecimen, sample):
+    """
+    Update Container using specific Biospecimen attributes.
+    Link Container to its associated biospecimen and sample
+    Validate Container
+    """
+    # Extract the parameters that uniquely identify a container
+    params = _get_container_identifier(biospecimen)
+    # Add remaining container attributes
+    params.update(
+        {
+            "biospecimen_id": biospecimen.kf_id,
+            "sample_id": sample.kf_id,
+            "volume_ul": biospecimen.volume_ul,
+            "external_id": biospecimen.external_aliquot_id,
+        }
+    )
+    # Set visibility params based on Biospecimen which represents both
+    # sample and containers
+    params.update(
+        _get_visibility_params(biospecimen)
+    )
+    try:
+        container = ContainerSchema(strict=True).load(
+            params, instance=current_container, partial=True
+        ).data
+    except ValidationError as e:
+        abort(400, 'could not update container: {}'.format(e.messages))
+
+    return container
+
+
+def _upsert_sample(biospecimen):
```

Comment:
Need to change this approach. Read, modify, write is an anti-pattern and doesn't work with concurrent requests. Use postgresql internal upsert (update on conflict)